src/ssh/cockpitsshrelay.c

```diff
@@ -1320,61 +1320,67 @@ send_auth_reply (CockpitSshData *data,
   return ret;
 }
 
-static void
+static gboolean
 parse_host (const gchar *host,
             gchar **hostname,
             gchar **username,
             guint *port)
 {
-  gchar *user_arg = NULL;
-  gchar *host_arg = NULL;
-  gchar *tmp = NULL;
-  gchar *end = NULL;
-
-  guint64 tmp_num;
-
-  gsize host_offset = 0;
-  gsize host_length = strlen (host);
-
-  tmp = strrchr (host, '@');
-  if (tmp)
-    {
-      if (tmp[0] != host[0])
+  GError *error = NULL;
+  g_autoptr (GRegex) regex = g_regex_new ("^"
+         "(?:(.+)@)?"         /* optional username */
+         "(?|"                /* one of... */
+           "\\[([^]@]+)\\]"     /* hostname in square brackets, no @ */
+           "(?::([0-9]+))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@:]+)"           /* hostname with no : or @ */
+           "(?::([0-9]+))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@]+)"            /* hostname with no @ but : (IPv6 address), and no port */
```

Comment:
You new case will also match a string like `hostname:ssh` (because of the failure to match the non-numeric port name in the previous iteration).  These aren't supported by `ssh(1)`, but then again, neither are numeric ports.

The thing that identifies an IPv6 address is that it always has at least two colons in it.  Maybe like `([^@]*:[^@]*:[^@]*)`.  That would match addresses like `2::1` `::1` `3::4:5:6%ifname` and others, along with `::`, which is a valid IPv6 address, but nothing that we'd ever want to SSH to.

We might also be okay to stop being so careful about the `@` thing.  The initial part, being greedy, should always consume every character up to the last `@`, as I understand it... although in the case that we *don't* get a match in th second part, it may decide to backtrack and try to get more creative...

Reply:
Or as just mentioned on IRC (writing it here so we don't forget): who cares?  We're arguing about the exact way in which this should fail to handle an invalid case.  It needs to fail anyway, and it will, later, when ssh rejects the hostname with a colon in it.

Reply:
Curiously, this is totally working with ssh:
```
Host p:foo
    Hostname piware.de
```

Then `ssh p:foo` actually works. So I agree that it's "good enough", and we kick the error handling can down the road towards ssh. The most important thing of this RE is that it does *not* mutilate an IPv6 address by misinterpreting a component of it as port, and reliably parses a user name and port. We don't need to go through great lengths of defining what a valid host name is, as that is very fungible.